Mathlib/NumberTheory/NumberField/Basic.lean

```diff
@@ -80,13 +80,8 @@ theorem isIntegral_of_mem_ringOfIntegers {K : Type _} [Field K] {x : K} (hx : x
     Polynomial.aeval_def, Subtype.coe_mk, hP]
 #align number_field.is_integral_of_mem_ring_of_integers NumberField.isIntegral_of_mem_ringOfIntegers
 
-/-- Given an algebra between two fields, create an algebra between their two rings of integers.
-
-For now, this is not an instance by default as it creates an equal-but-not-defeq diamond with
-`Algebra.id` when `K = L`. This is caused by `x = ⟨x, x.prop⟩` not being defeq on subtypes. This
-will likely change in Lean 4. -/
--- Porting note: check if this can be an instance now
-def ringOfIntegersAlgebra [Algebra K L] : Algebra (𝓞 K) (𝓞 L) :=
+/-- Given an algebra between two fields, create an algebra between their two rings of integers. -/
+instance inst_ringOfIntegersAlgebra [Algebra K L] : Algebra (𝓞 K) (𝓞 L) :=
```

Comment:
I think this instance is still bad because it creates a diamond for `𝓞 (𝓞 K)`, assuming that's legal. It's the same reason we don't have a `Algebra R[X] A[X]` instance.

Reply:
The "right" solution would be to copy the transitive closure mechanism we have for `Coe` and do the same for `Algebra`. (This is still on my todo list, I can get started with it at some point between a few months from now to a year from now.)

Reply:
`𝓞 (𝓞 K)` is not a thing, whether it's legal or not :-)

Reply:
> The "right" solution would be to copy the transitive closure mechanism we have for `Coe` and do the same for `Algebra`. (This is still on my todo list, I can get started with it at some point between a few months from now to a year from now.)

This sounds like a diamond disaster to me, we almost never want to have an `Algebra (F x) (F y)` instance because then `Algebra (F x) (F (F x))` can be found in two ways.

... of course, we already have two such instances, for pi types and pointwise set actions; but I'd still argue that we should not let any more worms out of this can.

Reply:
And yet somehow I encounter no coherence issues with `Coe`! Do I understand your point that this is because `Algebra` instances have a different shape than `Coe`?

Reply:
Here's a simple demo of the problem:
```lean
instance : Coe α (List α) := ⟨fun x => [x]⟩
instance [Coe α β] : Coe (List α) (List β) := ⟨fun x => x.map Coe.coe⟩

#eval (([1, 2, 3] : List Nat) : List (List Nat))

-- un-inlined the coercion, now the result is different!
def asList (a : α) : List α := a
#eval asList ([1, 2, 3] : List Nat)
```

Reply:
It sounds like I might be imagining that `Algebra K (𝓞 K)` exists?